consistency_check/src/bin/compressed_assets/main.rs

```diff
@@ -132,45 +121,45 @@ pub async fn main() {
 
     let assets_processed_clone = assets_processed.clone();
     let rate_clone = rate.clone();
-    let shutdown_token_clone = shutdown_token.clone();
-
     // update rate on the background
-    tokio::spawn(update_rate(shutdown_token_clone, assets_processed_clone, rate_clone));
+    usecase::executor::spawn(update_rate(
+        cancellation_token.child_token(),
+        assets_processed_clone,
+        rate_clone,
+    ));
 
     // write found problematic assets to the files
-    writers.spawn(async move {
-        let failed_checks_file =
-            File::create(FAILED_CHECKS_FILE).expect("Failed to create file for failed check trees");
-        let failed_proofs_file =
-            File::create(FAILED_PROOFS_FILE).expect("Failed to create file for failed proofs");
-
-        let mut failed_checks_wrt = csv::Writer::from_writer(failed_checks_file);
-        let mut failed_proofs_wrt = csv::Writer::from_writer(failed_proofs_file);
-
-        loop {
-            if let Err(e) = process_failed_checks(&mut failed_checks_wrt, &failed_check).await {
-                error!("Error writing failed checks: {}", e);
-            }
+    usecase::executor::spawn({
+        let cancellation_token = cancellation_token.child_token();
+        async move {
+            let failed_checks_file = File::create(FAILED_CHECKS_FILE)
+                .expect("Failed to create file for failed check trees");
+            let failed_proofs_file =
+                File::create(FAILED_PROOFS_FILE).expect("Failed to create file for failed proofs");
+
+            let mut failed_checks_wrt = csv::Writer::from_writer(failed_checks_file);
+            let mut failed_proofs_wrt = csv::Writer::from_writer(failed_proofs_file);
+
+            loop {
+                if let Err(e) = process_failed_checks(&mut failed_checks_wrt, &failed_check).await {
+                    error!("Error writing failed checks: {}", e);
+                }
 
-            if let Err(e) = process_failed_proofs(&mut failed_proofs_wrt, &failed_proofs).await {
-                error!("Error writing failed proofs: {}", e);
-            }
+                if let Err(e) = process_failed_proofs(&mut failed_proofs_wrt, &failed_proofs).await
+                {
+                    error!("Error writing failed proofs: {}", e);
+                }
 
-            if !shutdown_for_file_writer_rx.is_empty() {
-                break;
-            }
+                if cancellation_token.is_cancelled() {
```

Comment:
Does it mean that if cancellation is received during sleep or proccess\_failed\_proofs (which can take quite a long time I presume) we wait for their completion?

Reply:
yes, if we start the iteration it will complete at least once. that's the logic that was here before and i don't find it to be erroneous